functions/src/modules/users.functions.ts

```diff
@@ -0,0 +1,212 @@
+        const spreadSheetId = req.query.sheet ? req.query.sheet : null;
+        if (!spreadSheetId) {
+            res.status(badRequest.status).send(badRequest.res);
+        } else {
+            performUserSpreadSheetImport(spreadSheetId).then(response => {
+                res.send(response['res']);
+            });
+        }
+    }
+);
+
+// Helper functions importUserRegistrationData
+const performUserSpreadSheetImport = sheet => {
+    // Variables to populate laters (used for sending back a detailed response to the user)
+    let usersAdded = [];
+    let usersAlreadyExist = [];
+    return new Promise(async (resolve, reject) => {
+        const auth = await google.auth.getClient({
+            scopes: ['https://www.googleapis.com/auth/spreadsheets.readonly']
+        });
+
+        const spreadsheet = google.sheets({ version: 'v4', auth });
+        spreadsheet.spreadsheets.values.get(
+            { spreadsheetId: sheet, range: 'Registrations' },
+            (err, response) => {
+                if (err) {
+                    reject({
+                        status: 400,
+                        res: {
+                            status: false,
+                            message: 'Sheet not found',
+                            code: 'INVALID_SHEET_ID'
+                        }
+                    });
+                } else {
+                    var cache = [];
+                    var resp = JSON.parse(
+                        JSON.stringify(response, function(key, value) {
+                            if (typeof value === 'object' && value !== null) {
+                                if (cache.indexOf(value) !== -1) {
+                                    // Duplicate reference found
+                                    try {
+                                        // If this value does not reference a parent it can be deduped
+                                        return JSON.parse(
+                                            JSON.stringify(value)
+                                        );
+                                    } catch (error) {
+                                        // discard key if value cannot be deduped
+                                        return;
+                                    }
+                                }
+                                // Store value in our collection
+                                cache.push(value);
+                            }
+                            return value;
+                        })
+                    );
+
+                    // Take Actual Values and Field Names apart
+                    let sheetFields = resp.data.values[0];
+                    let sheetRows = resp.data.values.slice(1);
+
+                    // Filter out users (Only Allow the user Custom claim CR with value YES)
+                    const users = sheetRows.filter(
+                        user =>
+                            user[5] != '' &&

```

Comment:
This is how this library works, it doesn't return objects, instead, it returns an array of objects. I have to manually convert the response in the an object.

Reply:
It is stable I guess. Because order shouldn't change. The reason why it shouldn't change is the functiona cannot be dynamic at all. By that I mean is if you add a field in the Google Spreadsheets, let call it `userNickName`. So we have to tweak the code for makit the function support this field as well. Such as there is a field called `emailAddress`, which  we have to map to the Firebase field `email`, so the names aren't same. If we have to tweak the code on any addition, so we have to tweak the code if you decide to change the order of the Spreadsheet data, which I don't think is necessary at all.